### PR TITLE
engine/portforward: create/delete port forward objects in the API as well as on the state [ch12050]

### PR DESCRIPTION
**Pull request by maiamcc**
Hello @milas,

Please review the following commits I made in branch maiamcc/create-delete-pf-obj:

aa536e66e2e5d26d43d76192032e95222280353f (2021-06-03 17:49:23 -0400)
reorg tests

857602391204ea0d42940dbc870cf282ef2060f1 (2021-06-03 17:09:38 -0400)
more robust api creation/deletion tests

045603b4270bc3f30389c509c02c26c302f52e69 (2021-06-03 16:49:49 -0400)
some initial tests

d7c2e491421c6780ada6232eac32172e47a1c19d (2021-06-03 16:44:13 -0400)
api object creation/updates

09681635a1736b7181fd047f3f58c774a1e1b578 (2021-06-03 15:22:50 -0400)
portforward: subscriber creates/deletes api object in addition to sending action [ch12050]

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by maiamcc** [comment]
(Apologies for the size of the diff on the test file, i was mostly just polishing and streamlining into helper functions 😅  lmk if you need me to break this up)

File: internal/engine/portforward/subscriber_test.go [hunk 1]
```diff
@@ -467,30 +414,113 @@ func (f pfsFixture) WaitUntilDone() {
...
-	return len(pf.Spec.Forwards) == 1 && f.forwardWithHostMatches(pf.Spec.Forwards[0], expectedLocal, expectedContainer, expectedHost)
+func (f *pfsFixture) waitUntilStateAndAPIPortForward(msg string, name string, pfOK func(pf *PortForward) bool) {
+	f.T().Helper()
+	f.waitUntilStatePortForwards(msg, func(pfs map[string]*PortForward) bool {
+		if len(pfs) != 1 {
+			return false
+		}
+
+		var pf *PortForward
+		for _, portForward := range pfs { // there's only one thing in the map, we just don't know its key 🙃
```
Comment (milas) [inline internal/engine/portforward/subscriber_test.go:463]: Isn't its key the same as the API key below?

File: internal/engine/portforward/subscriber.go [hunk 2]
```diff
@@ -119,19 +127,65 @@ func (s *Subscriber) OnChange(ctx context.Context, st store.RStore, summary stor
...
+		st.Dispatch(store.NewErrorAction(fmt.Errorf(
+			"deleting PortForward from apiserver: %v", err)))
+		return
+	}
+	st.Dispatch(NewPortForwardDeleteAction(pf.Name))
+}
+
+// Create/update a PortForward API object, if necessary. Should be idempotent.
+func (s *Subscriber) upsertPF(ctx context.Context, st store.RStore, pf *PortForward) {
+	// TODO(maia): maybe simplify this code a bit by allowing create-on-update and just calling Update
```
Comment (milas) [inline internal/engine/portforward/subscriber.go:155]: I think this TODO can be dropped to avoid confusion later

**Review by milas — APPROVED** [review]

File: internal/engine/portforward/subscriber_test.go (hunk 1 above)
Comment (maiamcc) [reply internal/engine/portforward/subscriber_test.go:463]: ...maybe 😓

**Pull request merged**